### PR TITLE
Add gridscale builder

### PR DESCRIPTION
**Pull request by nvthongswansea**
The original branch https://github.com/gridscale/packer/tree/add-gridscale-builder (initialized by Ibrahim) has been changed to make gridscale builder testable and clean. The changes are:
- Break existing steps into many smaller and simpler steps.
- Add new (simple) steps.
- Update gsclient-go v3.
- Make it testable (using interfaces).
- Make it able to download ISO-Image from an URL.
- Fix artifacts.
- Support HCL2.
- Reduce amount of type assertions.
- Add comments.
- And some minor changes (go mod, etc.).

### CODE REVIEW

**Comment by bkircher** [comment]
@itakouna  :wave: Can you give your thoughts here?

**Review by bkircher — CHANGES_REQUESTED** [review]
I suspect files under `ansible/` should _not_ be part of what we want to push upstream. Correct?

**Comment by nvthongswansea** [comment]
I have no ideas. This is from Ibrahim's original branch.

**Comment by itakouna** [comment]
> I suspect files under `ansible/` should _not_ be part of what we want to push upstream. Correct?

yes, it should not! I added that for internal use!

**Comment by itakouna** [comment]
@nvthongswansea plz, remove `tests`folder

**Comment by nvthongswansea** [comment]
@bkircher @itakouna Done

File: builder/gridscale/builder.go [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+			client: client,
+			config: &b.config,
+			ui:     ui,
+		},
+		&stepLinkServerBootStorage{
+			client: client,
+			config: &b.config,
+			ui:     ui,
+		},
+		&stepCreateSecondaryStorage{
```
Comment (itakouna) [inline builder/gridscale/builder.go:63]: `SecondaryStorage` should be optional!

Comment (nvthongswansea) [reply builder/gridscale/builder.go:63]: Yes, it is optional in the step implementation.

Comment (nvthongswansea) [reply builder/gridscale/builder.go:63]: Check the step go file, you will see it is optional.

Comment (itakouna) [reply builder/gridscale/builder.go:63]: Ok.

**Comment by itakouna** [comment]
LGTM! we need to build and test it.

**Review by bkircher — APPROVED** [review]

**Comment by bkircher** [comment]
We should not merge this branch, because
 - we want to be able to rebase on upstream master
- and also this will be the base for another PR against upstream

**Comment by bkircher** [comment]
@itakouna 

> we need to build and test it

Yes. There are a couple of Jira issues where we add two downstream pipelines that build a (preliminary) template for us.

**Comment by bkircher** [comment]
FYI, I'll resurrect the branch merged here and force push. So you will need to update your remotes after that. I also will remove Ibranhim's branch (`add-gridscale-builder`) from this repo.

**Comment by bkircher** [comment]
Renamed the branch to add-gridscale-builder and rebased on upstream master.

**Pull request merged**